lib/scholar/spatial/distances.ex

```diff
@@ -0,0 +1,269 @@
+  """
+  @spec squared_euclidean(Nx.t(), Nx.t()) :: Nx.t()
+  defn squared_euclidean(x, y) do
+    assert_shape!(x, y)
+
+    x
+    |> Nx.subtract(y)
+    |> Nx.power(2)
+    |> Nx.sum()
+    |> Nx.as_type({:f, 32})
+  end
+
+  @doc """
+  Manhattan, taxicab, or l1 distance.
+
+  $$
+  D(x, y) = \\sum_i |x_i - y_i|
+  $$
+
+  ## Examples
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([3, 2])
+      iex> Scholar.Spatial.Distances.manhattan(x, y)
+      #Nx.Tensor<
+        f32
+        2.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2])
+      iex> Scholar.Spatial.Distances.manhattan(x, y)
+      #Nx.Tensor<
+        f32
+        0.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2, 3])
+      iex> Scholar.Spatial.Distances.manhattan(x, y)
+      ** (ArgumentError) expected input shapes to be equal, got {2} != {3}
+  """
+  @spec manhattan(Nx.t(), Nx.t()) :: Nx.t()
+  defn manhattan(x, y) do
+    assert_shape!(x, y)
+
+    x
+    |> Nx.subtract(y)
+    |> Nx.abs()
+    |> Nx.sum()
+    |> Nx.as_type({:f, 32})
+  end
+
+  @doc """
+  Chebyshev or l-infinity distance.
+
+  $$
+  D(x, y) = \\max_i |x_i - y_i|
+  $$
+
+  ## Examples
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([3, 2])
+      iex> Scholar.Spatial.Distances.chebyshev(x, y)
+      #Nx.Tensor<
+        f32
+        2.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2])
+      iex> Scholar.Spatial.Distances.chebyshev(x, y)
+      #Nx.Tensor<
+        f32
+        0.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2, 3])
+      iex> Scholar.Spatial.Distances.chebyshev(x, y)
+      ** (ArgumentError) expected input shapes to be equal, got {2} != {3}
+  """
+  @spec chebyshev(Nx.t(), Nx.t()) :: Nx.t()
+  defn chebyshev(x, y) do
+    assert_shape!(x, y)
+
+    x
+    |> Nx.subtract(y)
+    |> Nx.LinAlg.norm(ord: :inf)
+    |> Nx.as_type({:f, 32})
+  end
+
+  @doc """
+  Minkowski distance.
+
+  $$
+  D(x, y) = \\left(\\sum_i |x_i - y_i|^p\\right)^{\\frac{1}{p}}
+  $$
+
+  ## Examples
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([5, 2])
+      iex> Scholar.Spatial.Distances.minkowski(x, y)
+      #Nx.Tensor<
+        f32
+        4.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2])
+      iex> Scholar.Spatial.Distances.minkowski(x, y)
+      #Nx.Tensor<
+        f32
+        0.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2, 3])
+      iex> Scholar.Spatial.Distances.minkowski(x, y)
+      ** (ArgumentError) expected input shapes to be equal, got {2} != {3}
+  """
+  @spec minkowski(Nx.t(), Nx.t(), integer()) :: Nx.t()
+  defn minkowski(x, y, p \\ 2) do
+    assert_shape!(x, y)
+
+    x
+    |> Nx.subtract(y)
+    |> Nx.abs()
+    |> Nx.power(p)
+    |> Nx.sum()
+    |> Nx.power(1.0 / p)
+    |> Nx.as_type({:f, 32})
```

Comment:
This will downcast f64 to f32. Is it necessary? I think `Nx.power(..., float)` will always return a float already!